sdwan/resource_sdwan_device_template.go

```diff
@@ -0,0 +1,493 @@
+		for i := 0; i < len(gtcontList.Data().([]interface{})); i++ {
+			gtMap := make(map[string]interface{})
+
+			gtMap["template_id"] = stripQuotes(gtcontList.Index(i).S("templateId").String())
+
+			gtMap["template_type"] = stripQuotes(gtcontList.Index(i).S("templateType").String())
+
+			if len(gtcontList.Index(i).S("subTemplates").Children()) > 0 {
+				stcontList := gtcontList.Index(i).S("subTemplates")
+				stList := make([]interface{}, 0, 1)
+				for j := 0; j < len(stcontList.Data().([]interface{})); j++ {
+					stMap := make(map[string]string)
+
+					stMap["sub_template_id"] = stripQuotes(stcontList.Index(j).S("templateId").String())
+
+					stMap["sub_template_type"] = stripQuotes(stcontList.Index(j).S("templateType").String())
+
+					stList = append(stList, stMap)
+				}
+
+				gtMap["sub_templates"] = stList
+			} else {
+				gtMap["sub_templates"] = make([]interface{}, 0, 1)
+			}
+			gtList = append(gtList, gtMap)
+		}
+
+		d.Set("general_templates", gtList)
+	} else {
+		d.Set("general_templates", make([]interface{}, 0, 1))
+	}
+
+	d.SetId(stripQuotes(data.S("templateId").String()))
+
+	log.Println("[DEBUG] Ending Read method", d.Id())
+
+	return nil
+}
+
+func resourceSDWANDeviceTemplateUpdate(d *schema.ResourceData, m interface{}) error {
+	log.Println("[DEBUG] Begining Update method ", d.Id())
+
+	devTempID := d.Id()
+
+	sdwanClient := m.(*client.Client)
+
+	deviceTemplate := &models.DeviceTemplate{
+		TemplateName:        d.Get("template_name").(string),
+		TemplateDescription: d.Get("template_description").(string),
+		DeviceType:          d.Get("device_type").(string),
+		ConfigType:          d.Get("config_type").(string),
+		FactoryDefault:      d.Get("factory_default").(bool),
+	}
+
+	if deviceTemplate.ConfigType == "template" {
+		if connPref, ok := d.GetOk("connection_preference"); ok {
+			deviceTemplate.ConnectionPreference = connPref.(bool)
+		}
+
+		if connPrefReq, ok := d.GetOk("connection_preference_required"); ok {
+			deviceTemplate.ConnectionPreferenceRequired = connPrefReq.(bool)
+		}
+
+		if uidRange, ok := d.GetOk("feature_template_uid_range"); ok {
+			deviceTemplate.FeatureTemplateUidRange = interfaceToStrList(uidRange)
+		} else {
+			deviceTemplate.FeatureTemplateUidRange = make([]string, 0, 1)
+		}
+
+		if polID, ok := d.GetOk("policy_id"); ok {
+			deviceTemplate.PolicyId = polID.(string)
+		} else {
+			deviceTemplate.PolicyId = "{}"
+		}
+
+		if val, exists := d.GetOk("general_templates"); exists {
+			deviceTemplate.GeneralTemplates = expandGeneralTemplates(val.(*schema.Set).List())
+		}
+
+	} else {
+
+		if tconfig, ok := d.GetOk("template_configuration"); ok {
+			deviceTemplate.TemplateConfiguration = tconfig.(string)
+		}
+
+	}
+	dURL := fmt.Sprintf("/dataservice/template/device/%v", devTempID)
+	_, err := sdwanClient.Update(dURL, deviceTemplate)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(devTempID)
+
+	log.Println("[DEBUG] Ending Update method", d.Id())
+	return nil
```

Comment:
we have to return to READ method from the UPDATE as done in CREATE method
It should be ```return resourceSDWANDeviceTemplateRead(d, m)``` in place of ```return nil```